src/tools/tidy/src/error_codes_check.rs

```diff
@@ -0,0 +1,183 @@
+                    visit_dirs(&path, level + 1, cb)?;
+                }
+            }
+        } else {
+            cb(&entry);
+        }
+    }
+    Ok(())
+}
+
+fn open_file(file: &Path, cb: &mut dyn FnMut(BufReader<File>, String)) {
+    let f = match File::open(file) {
+        Ok(f) => f,
+        Err(e) => {
+            eprintln!("Failed to open {}: {}", file.display(), e);
+            return;
+        }
+    };
+    cb(BufReader::new(f), String::new())
+}
+
+fn extract_error_codes(
+    mut f: BufReader<File>,
+    mut buf: String,
+    error_codes: &mut HashMap<String, usize>,
+) {
+    let mut reached_no_explanation = false;
+    let mut last_error_code = None;
+    while let Ok(bytes_read) = f.read_line(&mut buf) {
+        if bytes_read == 0 {
+            return;
+        }
+        let s = buf.trim();
+        if s.starts_with('E') && s.ends_with(": r##\"") {
+            if let Some(err_code) = s.splitn(2, ':').next() {
+                let err_code = err_code.to_owned();
+                last_error_code = Some(err_code.clone());
+                if !error_codes.contains_key(&err_code) {
+                    error_codes.insert(err_code, 0);
+                }
+            }
+        } else if s.starts_with("```") && s.contains("compile_fail") && s.contains('E') {
+            if let Some(err_code) = s.splitn(2, 'E').skip(1).next() {
+                if let Some(err_code) = err_code.splitn(2, ',').next() {
+                    let nb = error_codes.entry(format!("E{}", err_code)).or_insert(0);
+                    *nb += 1;
+                }
+            }
+        } else if s == ";" {
+            reached_no_explanation = true;
+        } else if reached_no_explanation && s.starts_with('E') {
+            if let Some(err_code) = s.splitn(2, ',').next() {
+                let err_code = err_code.to_owned();
+                if !error_codes.contains_key(&err_code) { // this check should *never* fail!
+                    error_codes.insert(err_code, 0);
+                }
+            }
+        } else if s.starts_with("#### Note: this error code is no longer emitted by the compiler") {
+            if let Some(last) = last_error_code {
+                error_codes.get_mut(&last).map(|x| *x += 1);
+            }
+            last_error_code = None;
+        }
+        buf.clear();
+    }
+}
+
+fn extract_error_codes_from_tests(
+    mut f: BufReader<File>,
+    mut buf: String,
+    error_codes: &mut HashMap<String, usize>,
+) {
+    while let Ok(bytes_read) = f.read_line(&mut buf) {
+        if bytes_read == 0 {
+            return;
+        }
+        let s = buf.trim();
+        if s.starts_with("error[E") || s.starts_with("warning[E") {
+            if let Some(err_code) = s.splitn(2, ']').next() {
+                if let Some(err_code) = err_code.splitn(2, '[').skip(1).next() {
+                    let nb = error_codes.entry(err_code.to_owned()).or_insert(0);
+                    *nb += 1;
```

Comment:
Why keep a count instead of bool yes/no? We appear to only compare with 0...

Reply:
For now. I have in mind to use this to track error codes with no long error explanation. But for now a bool would be enough so I'll update it.